extensions/binding/org.eclipse.smarthome.binding.mqttgeneric/src/main/java/org/eclipse/smarthome/binding/mqttgeneric/handler/MqttThingHandler.java

```diff
@@ -0,0 +1,200 @@
+            } catch (MqttException e) {
+                logger.warn("Could not publish new value {} to broker", command.toString(), e);
+            }
+        }
+    }
+
+    @Override
+    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        if (bridgeStatusInfo.getStatus() == ThingStatus.ONLINE) {
+            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE);
+            // If the bridge is online, no handler null check needs to be performed
+            MqttBrokerConnectionHandler h = getBridgeHandler();
+            connection = h.getConnection();
+            for (ChannelConfig channelConfig : channelDataByChannelUID.values()) {
+                channelConfig.start(connection, this);
+            }
+
+        } else if (bridgeStatusInfo.getStatus() == ThingStatus.OFFLINE) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+            connection = null;
+        }
+    }
+
+    @Override
+    public void dispose() {
+        if (connection != null) {
+            for (ChannelConfig channelConfig : channelDataByChannelUID.values()) {
+                channelConfig.dispose();
+            }
+            connection = null;
+        }
+        channelDataByChannelUID.clear();
+    }
+
+    /**
+     * Return the bride handler.
+     * We make this an explicit function for test mocks.
+     */
+    protected MqttBrokerConnectionHandler getBridgeHandler() {
+        return (MqttBrokerConnectionHandler) getBridge().getHandler();
+    }
+
+    /**
+     * Return the bridge status.
+     * We make this an explicit function for test mocks.
+     */
+    protected ThingStatusInfo getBridgeStatus() {
+        Bridge b = getBridge();
+        if (b != null) {
+            return b.getStatusInfo();
+        } else {
+            return new ThingStatusInfo(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, null);
+        }
+    }
+
+    @Override
+    public void initialize() {
+        for (Channel channel : thing.getChannels()) {
+            ChannelConfig config = channel.getConfiguration().as(ChannelConfig.class);
+            config.channelUID = channel.getUID();
+
+            if (StringUtils.isNotBlank(config.transformationPattern)) {
+                int index = config.transformationPattern.indexOf(':');
+                if (index == -1) {
+                    throw new IllegalArgumentException(
+                            "The transformation pattern must consist of the type and the pattern separated by a colon");
+                }
+                String type = config.transformationPattern.substring(0, index).toUpperCase();
+                config.transformationPattern = config.transformationPattern.substring(index + 1);
+                config.transformationService = getTransformationService(type);
+                if (config.transformationService == null) {
+                    throw new IllegalArgumentException("The transformation type is unknown: " + type);
```

Comment:
This is not so much of a good idea... 

1. Don't keep references on OSGi services (which is essentially what you are doing here, although `TransformationHelper` unfortunately hides this fact pretty well)
2. The transformation service implementation which is configured for this channel _might_ not/currently be available but will be later on. If you let initialization fail because of this, the handler won't get a chance to recover from this.

Therefore, please only remember the transformation *name* and rather grab a fresh reference every time.